contracts/mystery-box/src/lib.rs

```diff
@@ -0,0 +1,142 @@
+#![no_std]
+
+multiversx_sc::imports!();
+multiversx_sc::derive_imports!();
+
+pub mod config;
+pub mod events;
+pub mod permissions_module;
+pub mod rewards;
+
+use config::{Reward, RewardType, MAX_PERCENTAGE, NFT_AMOUNT};
+
+#[multiversx_sc::contract]
+pub trait MysteryBox:
+    config::ConfigModule
+    + rewards::RewardsModule
+    + permissions_module::PermissionsModule
+    + events::EventsModule
+{
+    #[init]
+    fn init(&self, mystery_box_token_id: TokenIdentifier) {
+        require!(
+            mystery_box_token_id.is_valid_esdt_identifier(),
+            "Invalid token ID"
+        );
+        self.mystery_box_token_identifier()
+            .set(mystery_box_token_id);
+
+        let caller = self.blockchain().get_caller();
+        self.add_admin(&caller);
+    }
+
+    #[endpoint(createMysteryBox)]
+    fn create_mystery_box(
+        &self,
+        receiver: ManagedAddress,
+        mystery_box_setup: ManagedVec<Reward<Self::Api>>,
+    ) -> EsdtTokenPayment<Self::Api> {
+        let caller = self.blockchain().get_caller();
+        self.require_admin(&caller);
+
+        let mystery_box_token_id = self.mystery_box_token_identifier().get();
+
+        let mut accumulated_percentage = 0u64;
+
+        for mystery_box_reward in mystery_box_setup.into_iter() {
+            accumulated_percentage += mystery_box_reward.percentage_chance;
+            self.check_reward_validity(&mystery_box_reward);
+        }
+        require!(
+            accumulated_percentage == MAX_PERCENTAGE,
+            "The total percentage must be 100%"
+        );
+
+        let mystery_box_token_nonce = self.send().esdt_nft_create_compact(
+            &mystery_box_token_id,
+            &BigUint::from(NFT_AMOUNT),
+            &mystery_box_setup,
+        );
+
+        let user_payment = EsdtTokenPayment::new(
```

Comment:
There is no need to create new NFT every time. SFTs are ok, same information, no need to duplicate in separate NFTs.